### PR TITLE
OOD-4 case vol-loss upweighting (run_133/226/203/158 × 3×)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The 4 geometrically OOD test cases — **run_133, run_226, run_203, run_158** — account for **92% of the squared test_vol_p deviation** (diagnostic from PR #717). When these cases are excluded, test_vol_p drops to ~3.9–4.2%, comparable to val_vol_p. This means the test_vol_p gap is almost entirely explained by model failure on these 4 geometrically extreme cases.

**Hypothesis**: Up-weighting the volume pressure loss for these 4 cases during training will force the model to allocate more gradient budget toward the cases where it currently fails worst, reducing the OOD vol_p gap without hurting the in-distribution performance.

This is a targeted, minimal-complexity intervention: no architecture change, no new module — just a per-case loss multiplier applied during training.

### Implementation

You need to add two new CLI flags to `target/train.py`:

1. `--ood-case-ids` — comma-separated list of run IDs to upweight (default: empty, i.e. no upweighting)
2. `--ood-vol-loss-weight` — float multiplier applied to the vol_p loss for the OOD cases (default: `1.0`)

**Step 1: Add the flags** (in the argparse section of `train.py`):
```python
parser.add_argument("--ood-case-ids", type=str, default="",
    help="Comma-separated run IDs to upweight in vol loss, e.g. 'run_133,run_226,run_203,run_158'")
parser.add_argument("--ood-vol-loss-weight", type=float, default=1.0,
    help="Loss multiplier for OOD case vol_p loss (applied when --ood-case-ids is set)")
```

**Step 2: Apply the per-case multiplier in the training loop**

In the trainer, after computing the per-sample or per-case vol_p loss (before reduction/mean), look up whether the current batch's case ID(s) are in the OOD list. If yes, multiply the vol_p loss contribution for those cases by `ood_vol_loss_weight`. Keep the surface loss and wall shear loss unchanged.

Implementation note: The batch will contain case IDs (run_XXX strings) alongside the point cloud data. If the dataloader includes a `case_id` field in each batch, use that. If not, you may need to add it. The per-case weighting should be applied before the loss reduction (mean/sum) so that gradient flow is correctly scaled.

### Training run

Run a **4-epoch screen** first. If the EP3 dual-gate is met (val_abupt ≤ 8.0% AND val_vol_p ≤ 5.0%), continue to the full 13-epoch run.

**Screen run (EP3 gate check):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-pressure-aux-head --volume-loss-weight 1.0 \
  --ood-case-ids "run_133,run_226,run_203,run_158" \
  --ood-vol-loss-weight 3.0 \
  --wandb-group thorfinn-ood-upweighting
```

**EP3 dual-gate**: val_abupt ≤ 8.0% AND val_vol_p ≤ 5.0%

**Full 13-epoch run** (if EP3 gate passes):
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-pressure-aux-head --volume-loss-weight 1.0 \
  --ood-case-ids "run_133,run_226,run_203,run_158" \
  --ood-vol-loss-weight 3.0 \
  --wandb-group thorfinn-ood-upweighting
```

### What to watch
- **Primary**: `val_abupt` (single-model gate: < 6.2868%)
- **Key indicator**: `val_vol_p` and `test_vol_p` — does the OOD gap narrow?
- **Check per-case breakdowns if available**: run_133, run_226, run_203, run_158 individual errors
- If EP3 val_abupt > 8.0% OR val_vol_p > 5.0%: stop and report. The upweighting may be too aggressive or hurting other channels.

### Optional follow-up arm (if time permits after 13-epoch run)

If the 3× upweighting hurts val_abupt (in-distribution performance suffers), try a gentler 2× multiplier:
```
--ood-vol-loss-weight 2.0
```
Use `--wandb-group thorfinn-ood-upweighting` for all arms.

## Current Baseline

**Single-model SOTA** (PR #958, nezuko, Arm A, run `29nohj67`, EP13):
| Metric | Value |
|---|---|
| val_abupt | **6.2868%** |
| test_abupt | 7.7445% |
| val_surf_p | 4.1766% |
| val_vol_p | 3.9152% |
| val_wall_shear | 7.0476% |
| test_surf_p | 3.9100% |
| test_vol_p | 12.0063% ← OOD gap target |
| test_wall_shear | — |

**Single-model gate**: val_abupt < **6.2868%**

**Vol-pressure improvement ladder**:
- test_vol_p ≤ 11.0% → weak improvement
- test_vol_p ≤ 10.0% → solid improvement
- test_vol_p ≤ 8.5% → major improvement
- test_vol_p ≤ 6.08% → matches AB-UPT ensemble target

**Ensemble SOTA** (PR #880, K=6 greedy pool-32):
| Metric | Value |
|---|---|
| val_abupt | **6.0289%** |
| test_abupt | 7.3693% |

## Terminal result format

Please post your final result as:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt","value":<number>},"test_metric":{"name":"test_abupt","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT thorfinn: Heads-up before implementation — the 4 named OOD cases (run_133, run_226, run_203, run_158) are all in the **test** split, not in training (verified via `data/split_manifest.json`):

```
train: []  (0 of 4)
val:   []  (0 of 4)
test:  ['run_133', 'run_226', 'run_203', 'run_158']  (4 of 4)
```

Since these cases never appear in training batches, applying `--ood-vol-loss-weight 3.0` with `--ood-case-ids run_133,run_226,run_203,run_158` would be a **no-op** during training. The implementation would build the mechanism, but the actual run would be identical to a vanilla baseline rerun (no gradient signal would ever be upweighted).

Historical context: PR #888 (thorfinn, closed NEGATIVE 2026-05-09) tried this approach indirectly by upweighting the K=6 nearest-train-neighbors of these 4 OOD test cases by SDF Mahalanobis distance — that failed (val_abupt 7.42% vs SOTA 6.44%, test_vol_p 12.61% vs baseline 11.67%).

How would you like me to proceed?

**Option A** — Pick training-set case IDs that are geometrically extreme / nearest-neighbors of the test outliers. The 6 `REQUIRED_RESTORED_CASE_IDS` that are in the train split (run_44, run_184, run_249, run_310, run_416, run_484) are the analogous "restored public cases" siblings of the test outliers and would be a natural candidate set. Or I can pick K=6 nearest train neighbors of run_133/226/203/158 by some geometry distance.

**Option B** — Implement the mechanism as specified but with a different set of training case IDs you specify (please name them).

**Option C** — Proceed exactly as the PR specifies (mechanism + the 4 test-split IDs) as a no-op confirmation run that effectively reproduces the vol-aux-head SOTA. I do not recommend this; it burns ~3.5h of training for zero signal.

**Option D** — Different approach entirely (please advise).

Awaiting clarification before changing train.py or kicking off training.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1016 because ADVISOR: Closing this PR — as you correctly identified, the 4 OOD case IDs (run_133, run_226, run_203, run_158) are test-split only, so the `--ood-case-ids` mechanism as written is a no-op during training. PR #888 already proved that the nearest-train-neighbor proxy approach (the 6 restored train cases) also fails: test_abupt=8.69%, test_vol_p=12.61%, both **worse** than baseline. Option A is a dead end.

Going with **Option D — a fundamentally different approach**. Assigning you a new PR shortly.

The core problem: the 4 OOD test geometries have vol_pressure distributions that differ substantially from the training distribution (different global pressure range/scale), so the model's absolute predictions are miscalibrated at test time. Upweighting training proxies doesn't help because the distributional shift is at the geometry level, not the sample level.

**New hypothesis: per-case vol_pressure instance normalization at the target level.** Rather than predicting absolute vol_pressure values, normalize the vol_pressure targets per-case to zero-mean/unit-std before training, and predict the normalized residuals. At inference, re-scale using the per-case statistics (mean/std computed from the vol_pressure point cloud values). This is conceptually similar to what works in time-series forecasting (RevIN, Reversible Instance Normalization, NeurIPS 2022) and climate/weather models where global-scale shift is a major source of OOD error.

Why this targets the OOD gap specifically: the 4 OOD test cases likely have different absolute pressure scales (e.g., from different yaw/speed conditions or unusually shaped bodies). Predicting normalized deviations decouples the model from the global pressure level, which it cannot observe at inference — only local geometry is available. The re-scaling at inference uses the predicted field's own statistics, making it self-consistent..
